lib/icingadb/redisconnection.cpp

```diff
@@ -233,12 +253,57 @@ void RedisConnection::Connect(asio::yield_context& yc)
 	for (;;) {
 		try {
 			if (m_Path.IsEmpty()) {
-				Log(LogInformation, "IcingaDB")
-					<< "Trying to connect to Redis server (async) on host '" << m_Host << ":" << m_Port << "'";
+				if (m_TLSContext) {
+					Log(LogInformation, "IcingaDB")
+						<< "Trying to connect to Redis server (async, TLS) on host '" << m_Host << ":" << m_Port << "'";
+
+					auto conn (Shared<AsioTlsStream>::Make(m_Strand.context(), *m_TLSContext, m_Host));
+					auto& tlsConn (conn->next_layer());
+
+					icinga::Connect(conn->lowest_layer(), m_Host, Convert::ToString(m_Port), yc);
+					tlsConn.async_handshake(tlsConn.client, yc);
+
+					if (!m_CaPath.IsEmpty()) {
+						std::shared_ptr<X509> cert (tlsConn.GetPeerCertificate());
+
+						if (!cert) {
+							BOOST_THROW_EXCEPTION(std::runtime_error(
+								"Redis didn't present any TLS certificate."
+							));
+						}
+
+						if (!tlsConn.IsVerifyOK()) {
+							BOOST_THROW_EXCEPTION(std::runtime_error(
+								"TLS certificate validation failed: " + std::string(tlsConn.GetVerifyError())
+							));
+						}
 
-				auto conn (Shared<TcpConn>::Make(m_Strand.context()));
-				icinga::Connect(conn->next_layer(), m_Host, Convert::ToString(m_Port), yc);
-				m_TcpConn = std::move(conn);
+						String identity;
+
+						try {
+							identity = GetCertificateCN(cert);
+						} catch (const std::exception& ex) {
+							BOOST_THROW_EXCEPTION(std::runtime_error(
+								"Cannot get common name from TLS certificate: " + std::string(ex.what())
+							));
+						}
+
+						if (identity != m_Host) {
+							BOOST_THROW_EXCEPTION(std::runtime_error(
+								"Unexpected TLS certificate common name: '" + std::string(identity) + "'"
+							));
+						}
```

Comment:
I'm pretty sure OpenSSL provides better APIs for validating certificates than this. This completely ignores subject alternative names and wildcards.